### ISSUE TITLE
`dunai-examples`: Convert examples from SDL1 to SDL2

### ISSUE THREAD

**Issue by ivanperez-keera** [ivanperez-keera/dunai#446]
SDL 1 no longer compiles well with even relatively old versions of GHC, such as 8.8 and 8.10.

This prevents the CI job from completing because cabal cannot find an installation plan.

Porting the examples to SDL 2 would make them more relevant to current developers, and keep our CI infrastructure working well.

**Comment by ivanperez-keera** [comment]
Help is sought for this. There are four modules that use SDL:

```
dunai-examples/bouncingball/BouncingBall.hs
dunai-examples/classicfrp/GTest.hs
dunai-examples/classicfrp/Test.hs
dunai-examples/list/BouncingBall.hs
```

**Comment by mchav** [comment]
Thanks for the clear instructions. I'll start working on the bouncing ball example to start.

**Comment by ivanperez-keera** [comment]
Great! Feel free to send a PR when you have something you want me to take a look at, or just ask any questions you have.

**Comment by mchav** [comment]
The example `dunai-examples/list/BouncingBall.hs` doesn't seem to work.

1) You can't `cabal run` in the directory because it's not part of `cabal.project`.
2) It doesn't compile unless you remove the type signatures
```
BouncingBall.hs:31:15: error:
    • Couldn't match type ‘Identity’ with ‘IO’
      Expected: bearriver-0.14.11:FRP.BearRiver.InternalCore.SF
                  IO (Float, Float, Bool, Bool) [(Float, Float)]
        Actual: MSF
                  (ClockInfo Identity) (Float, Float, Bool, Bool) [(Float, Float)]
```
3) Even after all this the example freezes on mouse press.

Will fix this example as I port it.

**Comment by ivanperez-keera** [comment]
2. is related to https://github.com/ivanperez-keera/dunai/issues/454.

**Comment by ivanperez-keera** [comment]
I think sometimes the build job is not completing in the CI for this reason:

https://github.com/haskell-game/sdl2/issues/310

**Comment by mchav** [comment]
Thanks for addressing that.

Figured out the cause of 3. We have to use `notYet` [here](https://github.com/ivanperez-keera/dunai/blob/a7ed00e349a776d6e93cd57c368192edfffcd1c4/dunai-examples/list/BouncingBall.hs#L45) to delay the switching event. Is this the right intuition or am I missing something here?

**Comment by ivanperez-keera** [comment]
Hmm. Does that happen also with SDL1?

**Comment by mchav** [comment]
Yes. The example as is doesn't work. Both in SDL1 and when ported to SDL2. There seems to be an infinite loop when switching.

**Comment by ivanperez-keera** [comment]
Ok. Let me try and reproduce that.

In the meantime, I've sent a patch to the sdl2 project to get sdl2 compiling with older versions of GHC: https://github.com/haskell-game/sdl2/pull/311

**Comment by ivanperez-keera** [comment]
The `sdl2` patch has now been merged.

I don't know when the `sdl2` project plans to make a new release, but we can for now use the github repo to test that everything works.

**Comment by mchav** [comment]
Did you manage to reproduce 3?

**Comment by ivanperez-keera** [comment]
Just to document things here too: I was able to reproduce (3) and we created a separate issue to track that error (#457 ), which @mchav just fixed (#458) .

**Comment by mchav** [comment]
No worries. I'll keep the PRs handy though.
